### PR TITLE
New Ice.TypeIdAttribute in C# (replacement for ice_id/ice_staticId)

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR adds a new Ice.TypeIdAttribute "custom attribute" class in C# and modifies slice2cs to add this attribute to generated interfaces (skeletons and proxies), classes (mapped Slice classes) and user exception classes.

The Ice.TypeId custom attribute replaces the existing `ice_id` and `ice_staticId` methods on these classes. Nevertheless I kept `ice_id` on all exception classes for now, until a later refactoring of all exceptions. For user exception, `ice_id` is implemented using the Ice.TypeId attribute.

For servants, the Object<T> helper class caches the type IDs in a static readonly field, so there is no performance penalty, even though calls to ice_id, ice_ids and ice_isA should not be performance-critical. Likewise for generated classes and exceptions, the type IDs are cached in static readonly fields.

In terms of API, the mapping is as follows:

| Ice 3.7  | Ice 4.0 | Notes |
| ------------- | ------------- |------------- |
|` x.ice_id()`  |`Ice.TypeIdAttribute.GetTypeId(x.GetType())` | Custom attribute retrieval uses reflection |
| `X.ice_staticId()`  |  `Ice.TypeIdAttribute.GetTypeId(typeof(X))` | |
| `servant.ice_ids()` | `Ice.TypeIdAttribute.GetAllTypeIds(typeof(MappedInterface))`| for servants, the type IDs are listed in alphabetical order and include ::Ice::Object |
| N/A |  `Ice.TypeIdAttribute.GetAllTypeIds(typeof(MappedClass))` | the type IDs are listed from most derived to least derived (but without a common base class such as AnyClass)|

### CODE REVIEW

**Comment by pepone** [comment]
I'm not against using attributes but I think we should provide extension methods to access this

```
Ice.TypeIdAttribute.GetTypeId(typeof(X))
```
Could be

```
typeof(X).GetTypeId()
```

I still think the current API seems simpler here `x.ice_static_id()`. 

Overall I think we don't need any of these and we can just keep internal IDs for IceIsA implementation that can very well be a constant of the method, same for class and exception marshaling the instances can write its id but don't need to expose it.

**Comment by bernardnormier** [comment]
I moved the TypeIdAttribute static methods into a TypeExtensions class, as requested by pepone. The new method names are GetIceTypeId and GetAllIceTypeIds to avoid confusion with other type IDs.

File: csharp/src/Ice/TypeIdAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                    if (typeId != null)
+                    {
+                        result.Add(typeId);
+                        p = p.BaseType;
+                    }
+                    else
+                    {
+                        p = null;
+                    }
+                } while (p != null);
```
Comment (pepone) [inline csharp/src/Ice/TypeIdAttribute.cs]: Maybe a for loop can make this a bit simpler
```
for(Type? p = type; p != null; p = p.BaseType)
{
    if (GetIceTypeId(p) is string typeId)
    {
        result.Add(typeId);
    }
}
```

File: csharp/src/Ice/Exception.cs [hunk 2]
```diff
@@ -8,40 +8,29 @@
...
-            }
-            catch (Exception)
-            {
-                expected = "";
-                Debug.Assert(false);
-            }
-
-            throw new Ice.UnexpectedObjectException("expected element of type `" + expected + "' but received `" +
-                                                    type + "'", type, expected);
+            string typeId = Ice.TypeExtensions.GetIceTypeId(v.GetType())!;
```
Comment (pepone) [inline csharp/src/Ice/Exception.cs]: Why not call it as an extension method here?

Comment (bernardnormier) [reply csharp/src/Ice/Exception.cs]: It's in the IceInternal namespace and this code is going to be removed soon.

File: csharp/src/Ice/TypeIdAttribute.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/TypeIdAttribute.cs]: Fixed as suggested, except with a `break` in the event a type does not have an Ice type ID.

**Pull request merged**